// @vizia/format/src/lib/resolve-props.js
'use strict';

const defaultProps = {
    format: 'number',
    output: 'string',
    currency: 'USD',
    currencySymbol: null,
    mantissa: 2,
    trimMantissa: true,
    abbr: true,
    abbrThreshold: 9999.5,
    unitDisplay: 'narrow',
    bemBlock: 'viz-format',
    locale: 'en-US'
};

/**
 * Filters empty values from obj
 */
const omitEmptyProps = (props) => {
    return Object.entries(props)
        .filter(([, value]) => typeof value !== 'undefined' && value !== null)
        .reduce((result, [key, value]) => Object.assign(result, {[key]: value}), {});
};

/**
 * Resolves props based on input and defaults, applies overrides where appropriate.
 */
const resolveProps = (props) => {
    return [
        // Defaults
        () => defaultProps,
        // Input props
        () => omitEmptyProps(props),
        // Overrides (based on format)
        ({format, abbrThreshold}) => {
            switch (format) {
                case 'currency':
                    return {
                        trimMantissa: false
                    };
                case 'percent':
                    return {
                        abbrThreshold: abbrThreshold / 100
                    };
                case 'ordinal':
                    return {
                        mantissa: 0,
                        abbr: false
                    };
                default:
                    return {};
            }
        }
    ]
        .reduce((result, processor) => Object.assign({}, result, processor(result)), {});
};

module.exports = resolveProps;
